clang/lib/Driver/ToolChains/Clang.cpp

```diff
@@ -3298,7 +3298,18 @@ static void RenderFloatingPointOptions(const ToolChain &TC, const Driver &D,
   }
 
   // Handle __FINITE_MATH_ONLY__ similarly.
-  if (!HonorINFs && !HonorNaNs)
+  bool InfValues = true;
+  bool NanValues = true;
+  auto processArg = [&](const auto *Arg) {
+    if (StringRef(Arg->getValue()) == "-menable-no-nans")
+      NanValues = false;
+    if (StringRef(Arg->getValue()) == "-menable-no-infs")
+      InfValues = false;
+  };
+  for (auto *Arg : Args.filtered(options::OPT_Xclang))
+    processArg(Arg);
```

Comment:
Instead of filtering all arguments to cc1 and looping over all of them, can we use `Args.getLastArg()`? And can we skip that processing if we already know we're doing to pass `-ffinite-math-only` because of `(!HonorINFs && !HonorNaNs)`?

Reply:
I can't use `Args.getLastArg()` with arguments to `Xclang`. `Args.filtered()` needs to be used.

Reply:
Okay, thanks! We should still skip processing those args when we can though.

Reply:
That could work, I was thinking along these lines:
```
  // Explanation of why we need to do this dance goes here.
  auto NaNsAndInfs = [&] {
    bool InfValues = true;
    bool NanValues = true;
    auto processArg = [&](const auto *Arg) {
      if (StringRef(Arg->getValue()) == "-menable-no-nans")
        NanValues = false;
      if (StringRef(Arg->getValue()) == "-menable-no-infs")
        InfValues = false;
    };

    for (auto *Arg : Args.filtered(options::OPT_Xclang))
      processArg(Arg);

    return InfValues && NanValues;
  };

  if ((!HonorINFs && !HonorNaNs) || !NaNsAndInfs())
    CmdArgs.push_back("-ffinite-math-only");
```
I leave it to driver folks to say what they'd prefer if they have a strong preference.

Reply:
@mdtoguchi WDYT?

Reply:
No strong preference - singular addition of `-ffinite-math-only` is easier to maintain.

Reply:
I see. I could write something like this as a compromise?
`bool shouldAddFiniteMathOnly = false;`

`if (!HonorINFs && !HonorNaNs) {`
  `shouldAddFiniteMathOnly = true;`
`} else {`
  `bool InfValues = true;`
 `bool NanValues = true;`

  `for (const auto *Arg : Args.filtered(options::OPT_Xclang)) {`
    `StringRef ArgValue = Arg->getValue();`
    `if (ArgValue == "-menable-no-nans")`
      `NanValues = false;`
    `else if (ArgValue == "-menable-no-infs")`
      `InfValues = false;`
  `}`

 ` if (!NanValues && !InfValues)`
   `shouldAddFiniteMathOnly = true;`
`}`

`if (shouldAddFiniteMathOnly) {`
  `CmdArgs.push_back("-ffinite-math-only");`
`}`

Reply:
I'd be okay with that approach.